# Counting the number of Spell Bee puzzles

### Version 0: Theoretical maximum

In [6]:
from math import comb

def count_theoretical_max(alphabet_size=26, puzzle_size=7):
    # Total number of ways to choose 7 letters from 26
    letter_combinations = comb(alphabet_size, puzzle_size)
    
    # Each combination can have any of the 7 letters as center
    total_possibilities = letter_combinations * puzzle_size
    
    return total_possibilities

# Calculate and display result
result = count_theoretical_max()
print(f"Total possible Spell Bee puzzles: {result:,}")

Total possible Spell Bee puzzles: 4,604,600


### Version 1: Add rule that it must contain at least 1 vowel

In [7]:
def count_valid_puzzles(puzzle_size=7, min_vowels=1):
    VOWELS = set('AEIOU')
    CONSONANTS = set('BCDFGHJKLMNPQRSTVWXYZ')
    
    def has_vowel(letters):
        return bool(set(letters) & VOWELS)
    
    total = 0
    # Count combinations with at least 'min_vowels' vowels
    for vowel_count in range(min_vowels, min(puzzle_size, len(VOWELS)) + 1):
        consonant_count = puzzle_size - vowel_count
        if consonant_count <= len(CONSONANTS):
            # Calculate valid combinations
            vowel_combos = comb(len(VOWELS), vowel_count)
            consonant_combos = comb(len(CONSONANTS), consonant_count)
            total += vowel_combos * consonant_combos * puzzle_size # center letter choices

    return total

# Calculate and display result
result = count_valid_puzzles()
print(f"Total possible Spell Bee puzzles: {result:,}")

Total possible Spell Bee puzzles: 3,790,640


The above makes sense since we can also arrive at this number by the following: 

Total number of theoretical puzzles = count_theoretical_max(26,7) = 4,604,600

Total number of theoretical puzzles with only consonents = count_theoretical_max(21,7) = 813,960

The difference between the two will give us the number of theoretical puzzles with atleast one vowel, which yields 3,790,640.